[1.16.5] Crashing on launch.

Game is crashing on launch. java 8 used.

https://paste.ee/p/wvLGq


![image](https://user-images.githubusercontent.com/40701814/108051673-29a36700-7019-11eb-851e-f8fb016df81b.png)

Forge 36.0.15
FD 0.3.2

It is also worth pointing out that, by relying on static initializer, you are accidentally accessing your `RegistryObjects` too early. The following stacktrace are from the linked issue ticket above:

```
Caused by: java.lang.NullPointerException: Registry Object not present: farmersdelight:wild_cabbages
	at java.util.Objects.requireNonNull(Objects.java:290) ~[?:1.8.0_281] {}
	at net.minecraftforge.fml.RegistryObject.get(RegistryObject.java:120) ~[forge:?] {re:mixin,re:classloading}
	at vectorwing.farmersdelight.world.CropPatchGeneration.<clinit>(CropPatchGeneration.java:22) ~[farmersdelight:1.16.3-0.3.2] {re:classloading}
	... 14 more
```

This was in turn occured during

```
java.lang.ExceptionInInitializerError: null
	at vectorwing.farmersdelight.FarmersDelight.<init>(FarmersDelight.java:35) ~[farmersdelight:1.16.3-0.3.2] {re:classloading}
```

Of course your `ModBlocks` won't be ready when your mod is still constructing, and the registry events have not been dispatched yet.

~~Please consider migrating your `Feature<?>` registration, i.e. these~~

https://github.com/vectorwing/FarmersDelight/blob/8adb746d89e9aeeb4e963cbfe8204e6edf52a5ad/src/main/java/vectorwing/farmersdelight/world/CropPatchGeneration.java#L36-L49  

~~to `DeferredRegister` as well.~~ Edit: for `ConfiguredFeature` you apparently cannot use registry event, see below

--- a/src/main/java/vectorwing/farmersdelight/FarmersDelight.java
+++ b/src/main/java/vectorwing/farmersdelight/FarmersDelight.java
@@ -5,7 +5,6 @@ import net.minecraft.util.ResourceLocation;
 import net.minecraftforge.common.MinecraftForge;
 import net.minecraftforge.common.crafting.CraftingHelper;
 import net.minecraftforge.event.RegistryEvent;
-import net.minecraftforge.eventbus.api.EventPriority;
 import net.minecraftforge.eventbus.api.IEventBus;
 import net.minecraftforge.fml.ModLoadingContext;
 import net.minecraftforge.fml.common.Mod;
@@ -20,7 +19,6 @@ import vectorwing.farmersdelight.registry.*;
 import vectorwing.farmersdelight.setup.ClientEventHandler;
 import vectorwing.farmersdelight.setup.CommonEventHandler;
 import vectorwing.farmersdelight.setup.Configuration;
-import vectorwing.farmersdelight.world.CropPatchGeneration;
 
 @Mod(FarmersDelight.MODID)
 @Mod.EventBusSubscriber(modid = FarmersDelight.MODID, bus = Mod.EventBusSubscriber.Bus.MOD)
@@ -32,15 +30,14 @@ public class FarmersDelight
 	public static final FDItemGroup ITEM_GROUP = new FDItemGroup(FarmersDelight.MODID);
 
 	public FarmersDelight() {
-		FMLJavaModLoadingContext.get().getModEventBus().addListener(CommonEventHandler::init);
-		FMLJavaModLoadingContext.get().getModEventBus().addListener(ClientEventHandler::init);
-		FMLJavaModLoadingContext.get().getModEventBus().addGenericListener(IRecipeSerializer.class, this::registerRecipeSerializers);
-		MinecraftForge.EVENT_BUS.addListener(EventPriority.HIGH, CropPatchGeneration::onBiomeLoad);
+		final IEventBus modEventBus = FMLJavaModLoadingContext.get().getModEventBus();
+
+		modEventBus.addListener(CommonEventHandler::init);
+		modEventBus.addListener(ClientEventHandler::init);
+		modEventBus.addGenericListener(IRecipeSerializer.class, this::registerRecipeSerializers);
 
 		ModLoadingContext.get().registerConfig(ModConfig.Type.COMMON, Configuration.COMMON_CONFIG);
 		ModLoadingContext.get().registerConfig(ModConfig.Type.CLIENT, Configuration.CLIENT_CONFIG);
-
-		final IEventBus modEventBus = FMLJavaModLoadingContext.get().getModEventBus();
 
 		ModParticleTypes.PARTICLE_TYPES.register(modEventBus);
 		ModEnchantments.ENCHANTMENTS.register(modEventBus);

--- a/src/main/java/vectorwing/farmersdelight/setup/CommonEventHandler.java
+++ b/src/main/java/vectorwing/farmersdelight/setup/CommonEventHandler.java
@@ -18,11 +18,15 @@ import net.minecraft.potion.Effects;
 import net.minecraft.util.*;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.gen.GenerationStage;
 import net.minecraftforge.common.crafting.CraftingHelper;
+import net.minecraftforge.common.world.BiomeGenerationSettingsBuilder;
 import net.minecraftforge.event.LootTableLoadEvent;
 import net.minecraftforge.event.entity.living.LivingEntityUseItemEvent;
 import net.minecraftforge.event.entity.player.UseHoeEvent;
 import net.minecraftforge.event.village.VillagerTradesEvent;
+import net.minecraftforge.event.world.BiomeLoadingEvent;
 import net.minecraftforge.eventbus.api.Event;
 import net.minecraftforge.eventbus.api.SubscribeEvent;
 import net.minecraftforge.fml.common.Mod;
@@ -37,6 +41,7 @@ import vectorwing.farmersdelight.registry.ModEffects;
 import vectorwing.farmersdelight.registry.ModItems;
 import vectorwing.farmersdelight.tile.dispenser.CuttingBoardDispenseBehavior;
 import vectorwing.farmersdelight.utils.tags.ModTags;
+import vectorwing.farmersdelight.world.CropPatchGeneration;
 import vectorwing.farmersdelight.world.VillageStructures;
 
 import javax.annotation.ParametersAreNonnullByDefault;
@@ -58,8 +63,11 @@ public class CommonEventHandler
 	private static final String[] SCAVENGING_ENTITIES = new String[]{"cow", "pig", "chicken", "rabbit", "horse", "donkey", "mule", "llama", "spider", "hoglin", "shulker"};
 
 	public static void init(final FMLCommonSetupEvent event) {
-		registerCompostables(event);
-		registerDispenserBehaviors(event);
+		event.enqueueWork(() -> {
+			registerCompostables();
+			registerDispenserBehaviors();
+			CropPatchGeneration.registerConfiguredFeatures();
+		});
 
 		ModAdvancements.register();
 
@@ -73,83 +81,121 @@ public class CommonEventHandler
 		}
 	}
 
-	public static void registerDispenserBehaviors(final FMLCommonSetupEvent event) {
-		event.enqueueWork(() -> {
-			if (Configuration.DISPENSER_TOOLS_CUTTING_BOARD.get()) {
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.WOODEN_PICKAXE, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.WOODEN_AXE, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.WOODEN_SHOVEL, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.STONE_PICKAXE, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.STONE_AXE, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.STONE_SHOVEL, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.IRON_PICKAXE, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.IRON_AXE, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.IRON_SHOVEL, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.DIAMOND_PICKAXE, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.DIAMOND_AXE, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.DIAMOND_SHOVEL, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.GOLDEN_PICKAXE, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.GOLDEN_AXE, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.GOLDEN_SHOVEL, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.NETHERITE_PICKAXE, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.NETHERITE_AXE, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.NETHERITE_SHOVEL, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(Items.SHEARS, new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(ModItems.FLINT_KNIFE.get(), new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(ModItems.IRON_KNIFE.get(), new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(ModItems.DIAMOND_KNIFE.get(), new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(ModItems.GOLDEN_KNIFE.get(), new CuttingBoardDispenseBehavior());
-				CuttingBoardDispenseBehavior.registerBehaviour(ModItems.NETHERITE_KNIFE.get(), new CuttingBoardDispenseBehavior());
-			}
-		});
+	public static void registerDispenserBehaviors() {
+		if (Configuration.DISPENSER_TOOLS_CUTTING_BOARD.get()) {
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.WOODEN_PICKAXE, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.WOODEN_AXE, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.WOODEN_SHOVEL, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.STONE_PICKAXE, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.STONE_AXE, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.STONE_SHOVEL, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.IRON_PICKAXE, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.IRON_AXE, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.IRON_SHOVEL, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.DIAMOND_PICKAXE, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.DIAMOND_AXE, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.DIAMOND_SHOVEL, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.GOLDEN_PICKAXE, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.GOLDEN_AXE, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.GOLDEN_SHOVEL, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.NETHERITE_PICKAXE, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.NETHERITE_AXE, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.NETHERITE_SHOVEL, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(Items.SHEARS, new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(ModItems.FLINT_KNIFE.get(), new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(ModItems.IRON_KNIFE.get(), new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(ModItems.DIAMOND_KNIFE.get(), new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(ModItems.GOLDEN_KNIFE.get(), new CuttingBoardDispenseBehavior());
+			CuttingBoardDispenseBehavior.registerBehaviour(ModItems.NETHERITE_KNIFE.get(), new CuttingBoardDispenseBehavior());
+		}
 	}
 
-	public static void registerCompostables(final FMLCommonSetupEvent event) {
-		event.enqueueWork(() -> {
-			// 30% chance
-			ComposterBlock.CHANCES.put(ModItems.TREE_BARK.get(), 0.3F);
-			ComposterBlock.CHANCES.put(ModItems.STRAW.get(), 0.3F);
-			ComposterBlock.CHANCES.put(ModItems.CABBAGE_SEEDS.get(), 0.3F);
-			ComposterBlock.CHANCES.put(ModItems.TOMATO_SEEDS.get(), 0.3F);
-			ComposterBlock.CHANCES.put(ModItems.RICE.get(), 0.65F);
-			ComposterBlock.CHANCES.put(ModItems.RICE_PANICLE.get(), 0.65F);
+	public static void registerCompostables() {
+		// 30% chance
+		ComposterBlock.CHANCES.put(ModItems.TREE_BARK.get(), 0.3F);
+		ComposterBlock.CHANCES.put(ModItems.STRAW.get(), 0.3F);
+		ComposterBlock.CHANCES.put(ModItems.CABBAGE_SEEDS.get(), 0.3F);
+		ComposterBlock.CHANCES.put(ModItems.TOMATO_SEEDS.get(), 0.3F);
+		ComposterBlock.CHANCES.put(ModItems.RICE.get(), 0.65F);
+		ComposterBlock.CHANCES.put(ModItems.RICE_PANICLE.get(), 0.65F);
 
-			// 50% chance
-			ComposterBlock.CHANCES.put(ModItems.PUMPKIN_SLICE.get(), 0.65F);
-			ComposterBlock.CHANCES.put(ModItems.CABBAGE_LEAF.get(), 0.65F);
+		// 50% chance
+		ComposterBlock.CHANCES.put(ModItems.PUMPKIN_SLICE.get(), 0.65F);
+		ComposterBlock.CHANCES.put(ModItems.CABBAGE_LEAF.get(), 0.65F);
 
-			// 65% chance
-			ComposterBlock.CHANCES.put(ModItems.CABBAGE.get(), 0.65F);
-			ComposterBlock.CHANCES.put(ModItems.ONION.get(), 0.65F);
-			ComposterBlock.CHANCES.put(ModItems.TOMATO.get(), 0.65F);
-			ComposterBlock.CHANCES.put(ModItems.WILD_CABBAGES.get(), 0.65F);
-			ComposterBlock.CHANCES.put(ModItems.WILD_ONIONS.get(), 0.65F);
-			ComposterBlock.CHANCES.put(ModItems.WILD_TOMATOES.get(), 0.65F);
-			ComposterBlock.CHANCES.put(ModItems.WILD_CARROTS.get(), 0.65F);
-			ComposterBlock.CHANCES.put(ModItems.WILD_POTATOES.get(), 0.65F);
-			ComposterBlock.CHANCES.put(ModItems.WILD_BEETROOTS.get(), 0.65F);
-			ComposterBlock.CHANCES.put(ModItems.WILD_RICE.get(), 0.65F);
-			ComposterBlock.CHANCES.put(ModItems.PIE_CRUST.get(), 0.65F);
+		// 65% chance
+		ComposterBlock.CHANCES.put(ModItems.CABBAGE.get(), 0.65F);
+		ComposterBlock.CHANCES.put(ModItems.ONION.get(), 0.65F);
+		ComposterBlock.CHANCES.put(ModItems.TOMATO.get(), 0.65F);
+		ComposterBlock.CHANCES.put(ModItems.WILD_CABBAGES.get(), 0.65F);
+		ComposterBlock.CHANCES.put(ModItems.WILD_ONIONS.get(), 0.65F);
+		ComposterBlock.CHANCES.put(ModItems.WILD_TOMATOES.get(), 0.65F);
+		ComposterBlock.CHANCES.put(ModItems.WILD_CARROTS.get(), 0.65F);
+		ComposterBlock.CHANCES.put(ModItems.WILD_POTATOES.get(), 0.65F);
+		ComposterBlock.CHANCES.put(ModItems.WILD_BEETROOTS.get(), 0.65F);
+		ComposterBlock.CHANCES.put(ModItems.WILD_RICE.get(), 0.65F);
+		ComposterBlock.CHANCES.put(ModItems.PIE_CRUST.get(), 0.65F);
 
-			// 85% chance
-			ComposterBlock.CHANCES.put(ModItems.RICE_BALE.get(), 0.85F);
-			ComposterBlock.CHANCES.put(ModItems.SWEET_BERRY_COOKIE.get(), 0.85F);
-			ComposterBlock.CHANCES.put(ModItems.HONEY_COOKIE.get(), 0.85F);
-			ComposterBlock.CHANCES.put(ModItems.CAKE_SLICE.get(), 0.85F);
-			ComposterBlock.CHANCES.put(ModItems.APPLE_PIE_SLICE.get(), 0.85F);
-			ComposterBlock.CHANCES.put(ModItems.SWEET_BERRY_CHEESECAKE_SLICE.get(), 0.85F);
-			ComposterBlock.CHANCES.put(ModItems.CHOCOLATE_PIE_SLICE.get(), 0.85F);
-			ComposterBlock.CHANCES.put(ModItems.RAW_PASTA.get(), 0.85F);
+		// 85% chance
+		ComposterBlock.CHANCES.put(ModItems.RICE_BALE.get(), 0.85F);
+		ComposterBlock.CHANCES.put(ModItems.SWEET_BERRY_COOKIE.get(), 0.85F);
+		ComposterBlock.CHANCES.put(ModItems.HONEY_COOKIE.get(), 0.85F);
+		ComposterBlock.CHANCES.put(ModItems.CAKE_SLICE.get(), 0.85F);
+		ComposterBlock.CHANCES.put(ModItems.APPLE_PIE_SLICE.get(), 0.85F);
+		ComposterBlock.CHANCES.put(ModItems.SWEET_BERRY_CHEESECAKE_SLICE.get(), 0.85F);
+		ComposterBlock.CHANCES.put(ModItems.CHOCOLATE_PIE_SLICE.get(), 0.85F);
+		ComposterBlock.CHANCES.put(ModItems.RAW_PASTA.get(), 0.85F);
 
-			// 100% chance
-			ComposterBlock.CHANCES.put(ModItems.APPLE_PIE.get(), 1.0F);
-			ComposterBlock.CHANCES.put(ModItems.SWEET_BERRY_CHEESECAKE.get(), 1.0F);
-			ComposterBlock.CHANCES.put(ModItems.CHOCOLATE_PIE.get(), 1.0F);
-			ComposterBlock.CHANCES.put(ModItems.DUMPLINGS.get(), 1.0F);
-			ComposterBlock.CHANCES.put(ModItems.STUFFED_PUMPKIN.get(), 1.0F);
-			ComposterBlock.CHANCES.put(ModItems.BROWN_MUSHROOM_COLONY.get(), 1.0F);
-			ComposterBlock.CHANCES.put(ModItems.RED_MUSHROOM_COLONY.get(), 1.0F);
-		});
+		// 100% chance
+		ComposterBlock.CHANCES.put(ModItems.APPLE_PIE.get(), 1.0F);
+		ComposterBlock.CHANCES.put(ModItems.SWEET_BERRY_CHEESECAKE.get(), 1.0F);
+		ComposterBlock.CHANCES.put(ModItems.CHOCOLATE_PIE.get(), 1.0F);
+		ComposterBlock.CHANCES.put(ModItems.DUMPLINGS.get(), 1.0F);
+		ComposterBlock.CHANCES.put(ModItems.STUFFED_PUMPKIN.get(), 1.0F);
+		ComposterBlock.CHANCES.put(ModItems.BROWN_MUSHROOM_COLONY.get(), 1.0F);
+		ComposterBlock.CHANCES.put(ModItems.RED_MUSHROOM_COLONY.get(), 1.0F);
+	}
+
+	@SubscribeEvent
+	public static void onBiomeLoad(BiomeLoadingEvent event) {
+		BiomeGenerationSettingsBuilder builder = event.getGeneration();
+		Biome.Climate climate = event.getClimate();
+
+		if (event.getName().getPath().equals("beach")) {
+			if (Configuration.GENERATE_WILD_BEETROOTS.get()) {
+				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, CropPatchGeneration.PATCH_WILD_BEETROOTS);
+			}
+			if (Configuration.GENERATE_WILD_CABBAGES.get()) {
+				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, CropPatchGeneration.PATCH_WILD_CABBAGES);
+			}
+		}
+
+		if (event.getCategory().equals(Biome.Category.SWAMP) || event.getCategory().equals(Biome.Category.JUNGLE)) {
+			if (Configuration.GENERATE_WILD_RICE.get()) {
+				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, CropPatchGeneration.PATCH_WILD_RICE);
+			}
+		}
+
+		if (climate.temperature >= 1.0F) {
+			if (Configuration.GENERATE_WILD_TOMATOES.get()) {
+				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, CropPatchGeneration.PATCH_WILD_TOMATOES);
+			}
+		}
+
+		if (climate.temperature > 0.3F && climate.temperature < 1.0F) {
+			if (Configuration.GENERATE_WILD_CARROTS.get()) {
+				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, CropPatchGeneration.PATCH_WILD_CARROTS);
+			}
+			if (Configuration.GENERATE_WILD_ONIONS.get()) {
+				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, CropPatchGeneration.PATCH_WILD_ONIONS);
+			}
+		}
+
+		if (climate.temperature > 0.0F && climate.temperature <= 0.3F) {
+			if (Configuration.GENERATE_WILD_POTATOES.get()) {
+				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, CropPatchGeneration.PATCH_WILD_POTATOES);
+			}
+		}
 	}
 
 	@SubscribeEvent

--- a/src/main/java/vectorwing/farmersdelight/tile/CookingPotTileEntity.java
+++ b/src/main/java/vectorwing/farmersdelight/tile/CookingPotTileEntity.java
@@ -49,9 +49,9 @@ import java.util.Random;
 @ParametersAreNonnullByDefault
 public class CookingPotTileEntity extends TileEntity implements INamedContainerProvider, ITickableTileEntity, INameable
 {
-	public static final int MEAL_DISPLAY = 6;
-	public static final int CONTAINER_INPUT = 7;
-	public static final int FINAL_OUTPUT = 8;
+	public static final int MEAL_DISPLAY_SLOT = 6;
+	public static final int CONTAINER_SLOT = 7;
+	public static final int OUTPUT_SLOT = 8;
 	public static final int INVENTORY_SIZE = 9;
 
 	private ItemStackHandler itemHandler = createHandler();
@@ -167,7 +167,7 @@ public class CookingPotTileEntity extends TileEntity implements INamedContainerP
 
 		ItemStackHandler drops = new ItemStackHandler(INVENTORY_SIZE);
 		for (int i = 0; i < INVENTORY_SIZE; ++i) {
-			drops.setStackInSlot(i, i == MEAL_DISPLAY ? itemHandler.getStackInSlot(i) : ItemStack.EMPTY);
+			drops.setStackInSlot(i, i == MEAL_DISPLAY_SLOT ? itemHandler.getStackInSlot(i) : ItemStack.EMPTY);
 		}
 		if (this.customName != null) {
 			compound.putString("CustomName", ITextComponent.Serializer.toJson(this.customName));
@@ -208,7 +208,7 @@ public class CookingPotTileEntity extends TileEntity implements INamedContainerP
 				if (!this.doesMealHaveContainer(meal)) {
 					this.moveMealToOutput();
 					dirty = true;
-				} else if (!itemHandler.getStackInSlot(CONTAINER_INPUT).isEmpty()) {
+				} else if (!itemHandler.getStackInSlot(CONTAINER_SLOT).isEmpty()) {
 					this.useStoredContainersOnMeal();
 					dirty = true;
 				}
@@ -242,7 +242,7 @@ public class CookingPotTileEntity extends TileEntity implements INamedContainerP
 	}
 
 	private boolean hasInput() {
-		for (int i = 0; i < MEAL_DISPLAY; ++i) {
+		for (int i = 0; i < MEAL_DISPLAY_SLOT; ++i) {
 			if (!itemHandler.getStackInSlot(i).isEmpty()) return true;
 		}
 		return false;
@@ -254,12 +254,12 @@ public class CookingPotTileEntity extends TileEntity implements INamedContainerP
 			if (recipeOutput.isEmpty()) {
 				return false;
 			} else {
-				ItemStack currentOutput = itemHandler.getStackInSlot(MEAL_DISPLAY);
+				ItemStack currentOutput = itemHandler.getStackInSlot(MEAL_DISPLAY_SLOT);
 				if (currentOutput.isEmpty()) {
 					return true;
 				} else if (!currentOutput.isItemEqual(recipeOutput)) {
 					return false;
-				} else if (currentOutput.getCount() + recipeOutput.getCount() <= itemHandler.getSlotLimit(MEAL_DISPLAY)) {
+				} else if (currentOutput.getCount() + recipeOutput.getCount() <= itemHandler.getSlotLimit(MEAL_DISPLAY_SLOT)) {
 					return true;
 				} else {
 					return currentOutput.getCount() + recipeOutput.getCount() <= recipeOutput.getMaxStackSize();
@@ -274,14 +274,14 @@ public class CookingPotTileEntity extends TileEntity implements INamedContainerP
 		if (recipe != null && this.canCook(recipe)) {
 			this.container = this.getRecipeContainer();
 			ItemStack recipeOutput = recipe.getRecipeOutput();
-			ItemStack currentOutput = itemHandler.getStackInSlot(MEAL_DISPLAY);
+			ItemStack currentOutput = itemHandler.getStackInSlot(MEAL_DISPLAY_SLOT);
 			if (currentOutput.isEmpty()) {
-				itemHandler.setStackInSlot(MEAL_DISPLAY, recipeOutput.copy());
+				itemHandler.setStackInSlot(MEAL_DISPLAY_SLOT, recipeOutput.copy());
 			} else if (currentOutput.getItem() == recipeOutput.getItem()) {
 				currentOutput.grow(recipeOutput.getCount());
 			}
 		}
-		for (int i = 0; i < MEAL_DISPLAY; ++i) {
+		for (int i = 0; i < MEAL_DISPLAY_SLOT; ++i) {
 			if (itemHandler.getStackInSlot(i).hasContainerItem()) {
 				Direction direction = this.getBlockState().get(CookingPotBlock.FACING).rotateYCCW();
 				double dropX = pos.getX() + 0.5 + (direction.getXOffset() * 0.25);
@@ -317,7 +317,7 @@ public class CookingPotTileEntity extends TileEntity implements INamedContainerP
 	}
 
 	public ItemStack getMeal() {
-		return itemHandler.getStackInSlot(MEAL_DISPLAY);
+		return itemHandler.getStackInSlot(MEAL_DISPLAY_SLOT);
 	}
 
 	// ======== CUSTOM THINGS ========
@@ -344,7 +344,7 @@ public class CookingPotTileEntity extends TileEntity implements INamedContainerP
 	public NonNullList<ItemStack> getDroppableInventory() {
 		NonNullList<ItemStack> drops = NonNullList.create();
 		for (int i = 0; i < INVENTORY_SIZE; ++i) {
-			drops.add(i == MEAL_DISPLAY ? ItemStack.EMPTY : itemHandler.getStackInSlot(i));
+			drops.add(i == MEAL_DISPLAY_SLOT ? ItemStack.EMPTY : itemHandler.getStackInSlot(i));
 		}
 		return drops;
 	}
@@ -354,11 +354,11 @@ public class CookingPotTileEntity extends TileEntity implements INamedContainerP
 	 * Does NOT check if the meal has a container; this is done on tick.
 	 */
 	private void moveMealToOutput() {
-		ItemStack mealDisplay = itemHandler.getStackInSlot(MEAL_DISPLAY);
-		ItemStack finalOutput = itemHandler.getStackInSlot(FINAL_OUTPUT);
+		ItemStack mealDisplay = itemHandler.getStackInSlot(MEAL_DISPLAY_SLOT);
+		ItemStack finalOutput = itemHandler.getStackInSlot(OUTPUT_SLOT);
 		int mealCount = Math.min(mealDisplay.getCount(), mealDisplay.getMaxStackSize() - finalOutput.getCount());
 		if (finalOutput.isEmpty()) {
-			itemHandler.setStackInSlot(FINAL_OUTPUT, mealDisplay.split(mealCount));
+			itemHandler.setStackInSlot(OUTPUT_SLOT, mealDisplay.split(mealCount));
 		} else if (finalOutput.getItem() == mealDisplay.getItem()) {
 			mealDisplay.shrink(mealCount);
 			finalOutput.grow(mealCount);
@@ -370,16 +370,16 @@ public class CookingPotTileEntity extends TileEntity implements INamedContainerP
 	 * If input and meal containers don't match, nothing happens.
 	 */
 	private void useStoredContainersOnMeal() {
-		ItemStack mealDisplay = itemHandler.getStackInSlot(MEAL_DISPLAY);
-		ItemStack containerInput = itemHandler.getStackInSlot(CONTAINER_INPUT);
-		ItemStack finalOutput = itemHandler.getStackInSlot(FINAL_OUTPUT);
+		ItemStack mealDisplay = itemHandler.getStackInSlot(MEAL_DISPLAY_SLOT);
+		ItemStack containerInput = itemHandler.getStackInSlot(CONTAINER_SLOT);
+		ItemStack finalOutput = itemHandler.getStackInSlot(OUTPUT_SLOT);
 
 		if (isContainerValid(containerInput) && finalOutput.getCount() < finalOutput.getMaxStackSize()) {
 			int smallerStack = Math.min(mealDisplay.getCount(), containerInput.getCount());
 			int mealCount = Math.min(smallerStack, mealDisplay.getMaxStackSize() - finalOutput.getCount());
 			if (finalOutput.isEmpty()) {
 				containerInput.shrink(mealCount);
-				itemHandler.setStackInSlot(FINAL_OUTPUT, mealDisplay.split(mealCount));
+				itemHandler.setStackInSlot(OUTPUT_SLOT, mealDisplay.split(mealCount));
 			} else if (finalOutput.getItem() == mealDisplay.getItem()) {
 				mealDisplay.shrink(mealCount);
 				containerInput.shrink(mealCount);
@@ -452,7 +452,7 @@ public class CookingPotTileEntity extends TileEntity implements INamedContainerP
 		{
 			@Override
 			protected void onContentsChanged(int slot) {
-				if (slot >= 0 && slot < MEAL_DISPLAY) {
+				if (slot >= 0 && slot < MEAL_DISPLAY_SLOT) {
 					cookTimeTotal = getCookTime();
 					inventoryChanged();
 				}

--- a/src/main/java/vectorwing/farmersdelight/world/CropPatchGeneration.java
+++ b/src/main/java/vectorwing/farmersdelight/world/CropPatchGeneration.java
@@ -4,14 +4,9 @@ import com.google.common.collect.ImmutableSet;
 import net.minecraft.block.Blocks;
 import net.minecraft.util.registry.Registry;
 import net.minecraft.util.registry.WorldGenRegistries;
-import net.minecraft.world.biome.Biome;
-import net.minecraft.world.gen.GenerationStage;
 import net.minecraft.world.gen.blockplacer.SimpleBlockPlacer;
 import net.minecraft.world.gen.blockstateprovider.SimpleBlockStateProvider;
 import net.minecraft.world.gen.feature.*;
-import net.minecraftforge.common.world.BiomeGenerationSettingsBuilder;
-import net.minecraftforge.event.world.BiomeLoadingEvent;
-import net.minecraftforge.eventbus.api.SubscribeEvent;
 import vectorwing.farmersdelight.registry.ModBiomeFeatures;
 import vectorwing.farmersdelight.registry.ModBlocks;
 import vectorwing.farmersdelight.setup.Configuration;
@@ -33,64 +28,32 @@ public class CropPatchGeneration
 	public static final BlockClusterFeatureConfig RICE_PATCH_CONFIG = (new BlockClusterFeatureConfig.Builder(
 			new SimpleBlockStateProvider(ModBlocks.WILD_RICE.get().getDefaultState()), new SimpleBlockPlacer())).tries(64).xSpread(4).zSpread(4).whitelist(ImmutableSet.of(Blocks.DIRT.getBlock())).func_227317_b_().build();
 
-	public static final ConfiguredFeature<?, ?> PATCH_WILD_CABBAGES = register("patch_wild_cabbages", Feature.RANDOM_PATCH.withConfiguration(CABBAGE_PATCH_CONFIG)
-			.withPlacement(Features.Placements.HEIGHTMAP_PLACEMENT).chance(Configuration.CHANCE_WILD_CABBAGES.get()));
-	public static final ConfiguredFeature<?, ?> PATCH_WILD_ONIONS = register("patch_wild_onions", Feature.RANDOM_PATCH.withConfiguration(ONION_PATCH_CONFIG)
-			.withPlacement(Features.Placements.PATCH_PLACEMENT).chance(Configuration.CHANCE_WILD_ONIONS.get()));
-	public static final ConfiguredFeature<?, ?> PATCH_WILD_TOMATOES = register("patch_wild_tomatoes", Feature.RANDOM_PATCH.withConfiguration(TOMATO_PATCH_CONFIG)
-			.withPlacement(Features.Placements.PATCH_PLACEMENT).chance(Configuration.CHANCE_WILD_TOMATOES.get()));
-	public static final ConfiguredFeature<?, ?> PATCH_WILD_CARROTS = register("patch_wild_carrots", Feature.RANDOM_PATCH.withConfiguration(CARROT_PATCH_CONFIG)
-			.withPlacement(Features.Placements.PATCH_PLACEMENT).chance(Configuration.CHANCE_WILD_CARROTS.get()));
-	public static final ConfiguredFeature<?, ?> PATCH_WILD_POTATOES = register("patch_wild_potatoes", Feature.RANDOM_PATCH.withConfiguration(POTATO_PATCH_CONFIG)
-			.withPlacement(Features.Placements.PATCH_PLACEMENT).chance(Configuration.CHANCE_WILD_POTATOES.get()));
-	public static final ConfiguredFeature<?, ?> PATCH_WILD_BEETROOTS = register("patch_wild_beetroots", Feature.RANDOM_PATCH.withConfiguration(BEETROOT_PATCH_CONFIG)
-			.withPlacement(Features.Placements.HEIGHTMAP_PLACEMENT).chance(Configuration.CHANCE_WILD_BEETROOTS.get()));
-	public static final ConfiguredFeature<?, ?> PATCH_WILD_RICE = register("patch_wild_rice", ModBiomeFeatures.RICE.get().withConfiguration(RICE_PATCH_CONFIG)
-			.withPlacement(Features.Placements.HEIGHTMAP_PLACEMENT).chance(Configuration.CHANCE_WILD_RICE.get()));
+	public static final ConfiguredFeature<?, ?> PATCH_WILD_CABBAGES = Feature.RANDOM_PATCH.withConfiguration(CABBAGE_PATCH_CONFIG)
+			.withPlacement(Features.Placements.HEIGHTMAP_PLACEMENT).chance(Configuration.CHANCE_WILD_CABBAGES.get());
+	public static final ConfiguredFeature<?, ?> PATCH_WILD_ONIONS = Feature.RANDOM_PATCH.withConfiguration(ONION_PATCH_CONFIG)
+			.withPlacement(Features.Placements.PATCH_PLACEMENT).chance(Configuration.CHANCE_WILD_ONIONS.get());
+	public static final ConfiguredFeature<?, ?> PATCH_WILD_TOMATOES = Feature.RANDOM_PATCH.withConfiguration(TOMATO_PATCH_CONFIG)
+			.withPlacement(Features.Placements.PATCH_PLACEMENT).chance(Configuration.CHANCE_WILD_TOMATOES.get());
+	public static final ConfiguredFeature<?, ?> PATCH_WILD_CARROTS = Feature.RANDOM_PATCH.withConfiguration(CARROT_PATCH_CONFIG)
+			.withPlacement(Features.Placements.PATCH_PLACEMENT).chance(Configuration.CHANCE_WILD_CARROTS.get());
+	public static final ConfiguredFeature<?, ?> PATCH_WILD_POTATOES = Feature.RANDOM_PATCH.withConfiguration(POTATO_PATCH_CONFIG)
+			.withPlacement(Features.Placements.PATCH_PLACEMENT).chance(Configuration.CHANCE_WILD_POTATOES.get());
+	public static final ConfiguredFeature<?, ?> PATCH_WILD_BEETROOTS = Feature.RANDOM_PATCH.withConfiguration(BEETROOT_PATCH_CONFIG)
+			.withPlacement(Features.Placements.HEIGHTMAP_PLACEMENT).chance(Configuration.CHANCE_WILD_BEETROOTS.get());
+	public static final ConfiguredFeature<?, ?> PATCH_WILD_RICE = ModBiomeFeatures.RICE.get().withConfiguration(RICE_PATCH_CONFIG)
+			.withPlacement(Features.Placements.HEIGHTMAP_PLACEMENT).chance(Configuration.CHANCE_WILD_RICE.get());
 
 	private static <FC extends IFeatureConfig> ConfiguredFeature<FC, ?> register(String key, ConfiguredFeature<FC, ?> configuredFeature) {
 		return Registry.register(WorldGenRegistries.CONFIGURED_FEATURE, key, configuredFeature);
 	}
 
-	@SubscribeEvent
-	public static void onBiomeLoad(BiomeLoadingEvent event) {
-		BiomeGenerationSettingsBuilder builder = event.getGeneration();
-		Biome.Climate climate = event.getClimate();
-
-		if (event.getName().getPath().equals("beach")) {
-			if (Configuration.GENERATE_WILD_BEETROOTS.get()) {
-				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, PATCH_WILD_BEETROOTS);
-			}
-			if (Configuration.GENERATE_WILD_CABBAGES.get()) {
-				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, PATCH_WILD_CABBAGES);
-			}
-		}
-
-		if (event.getCategory().equals(Biome.Category.SWAMP) || event.getCategory().equals(Biome.Category.JUNGLE)) {
-			if (Configuration.GENERATE_WILD_RICE.get()) {
-				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, PATCH_WILD_RICE);
-			}
-		}
-
-		if (climate.temperature >= 1.0F) {
-			if (Configuration.GENERATE_WILD_TOMATOES.get()) {
-				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, PATCH_WILD_TOMATOES);
-			}
-		}
-
-		if (climate.temperature > 0.3F && climate.temperature < 1.0F) {
-			if (Configuration.GENERATE_WILD_CARROTS.get()) {
-				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, PATCH_WILD_CARROTS);
-			}
-			if (Configuration.GENERATE_WILD_ONIONS.get()) {
-				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, PATCH_WILD_ONIONS);
-			}
-		}
-
-		if (climate.temperature > 0.0F && climate.temperature <= 0.3F) {
-			if (Configuration.GENERATE_WILD_POTATOES.get()) {
-				builder.withFeature(GenerationStage.Decoration.VEGETAL_DECORATION, PATCH_WILD_POTATOES);
-			}
-		}
+	public static void registerConfiguredFeatures() {
+		register("patch_wild_cabbages", PATCH_WILD_CABBAGES);
+		register("patch_wild_onions", PATCH_WILD_ONIONS);
+		register("patch_wild_tomatoes", PATCH_WILD_TOMATOES);
+		register("patch_wild_carrots", PATCH_WILD_CARROTS);
+		register("patch_wild_potatoes", PATCH_WILD_POTATOES);
+		register("patch_wild_beetroots", PATCH_WILD_BEETROOTS);
+		register("patch_wild_rice", PATCH_WILD_RICE);
 	}
 }

--- a/src/main/java/vectorwing/farmersdelight/world/features/RiceCropFeature.java
+++ b/src/main/java/vectorwing/farmersdelight/world/features/RiceCropFeature.java
@@ -21,7 +21,6 @@ public class RiceCropFeature extends Feature<BlockClusterFeatureConfig>
 		super(configFactoryIn);
 	}
 
-
 	@Override
 	public boolean generate(ISeedReader worldIn, ChunkGenerator generator, Random rand, BlockPos pos, BlockClusterFeatureConfig config) {
 		BlockPos blockpos = worldIn.getHeight(Heightmap.Type.OCEAN_FLOOR_WG, pos);

